src/PolyType.Roslyn/SourceWriter.cs

```diff
@@ -187,11 +262,148 @@ private static ReadOnlySpan<char> GetNextLine(ref ReadOnlySpan<char> remainingTe
         return next;
     }
 
-    private unsafe void AppendSpan(ReadOnlySpan<char> span)
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private void Append(char value)
     {
-        fixed (char* ptr = span)
+        EnsureCapacity(_length + 1);
+        _buffer[_length++] = value;
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private void Append(ReadOnlySpan<char> span)
+    {
+        if (span.IsEmpty)
         {
-            _sb.Append(ptr, span.Length);
+            return;
         }
+
+        EnsureCapacity(_length + span.Length);
+        span.CopyTo(_buffer.AsSpan(_length));
+        _length += span.Length;
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private void Append(string text)
+    {
+        if (string.IsNullOrEmpty(text))
+        {
+            return;
+        }
+
+        EnsureCapacity(_length + text.Length);
+        text.AsSpan().CopyTo(_buffer.AsSpan(_length));
+        _length += text.Length;
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private void AppendLine()
+    {
+        EnsureCapacity(_length + _newLine.Length);
+        _newLine.AsSpan().CopyTo(_buffer.AsSpan(_length));
+        _length += _newLine.Length;
+    }
+
+    private void EnsureCapacity(int requiredCapacity)
+    {
+        if (requiredCapacity <= _buffer.Length)
+        {
+            return;
+        }
+
+        int newCapacity = Math.Max(requiredCapacity, _buffer.Length * 2);
+        char[] newBuffer = ArrayPool<char>.Shared.Rent(newCapacity);
+        _buffer.AsSpan(0, _length).CopyTo(newBuffer);
+        ArrayPool<char>.Shared.Return(_buffer);
+        _buffer = newBuffer;
+    }
+
+    /// <summary>
+    /// Disposes the SourceWriter and returns buffers to the pool.
+    /// </summary>
+    public void Dispose()
+    {
+        if (_disposed)
+        {
+            return;
+        }
+
+        ArrayPool<char>.Shared.Return(_buffer);
+        _buffer = null!;
+        _length = 0;
+        _disposed = true;
+    }
+
+#if NET6_0_OR_GREATER
```

Comment:
You can remove the build conditionals, as the InterpolatedStringHandler attribute is polyfilled by PolySharp.